src/software/estop/threaded_estop_reader.cpp

```diff
@@ -0,0 +1,89 @@
+
+#include "threaded_estop_reader.h"
+
+#include <boost/bind/bind.hpp>
+#include <boost/function.hpp>
+#include <cstdio>
+#include <cstring>
+#include <thread>
+#include <utility>
+
+#include "software/logger/logger.h"
+
+
+ThreadedEstopReader::ThreadedEstopReader(int startup_ms, int interval_ms,
+                                         std::unique_ptr<UartCommunication> uart_reader)
+    : startup_ms(startup_ms),
+      interval_ms(interval_ms),
+      timer(io_service, this->interval_ms),
+      estop_msg(0),
+      uart_reader(std::move(uart_reader)),
+      estop_state(EstopState::STOP)
+{
+    estop_thread = std::thread(boost::bind(&ThreadedEstopReader::continousRead, this));
+}
+
+void ThreadedEstopReader::continousRead()
+{
+    timer.async_wait(
+        boost::bind(&ThreadedEstopReader::tick, this, boost::asio::placeholders::error));
+    io_service.run();
+}
+
+EstopState ThreadedEstopReader::getEstopState()
+{
+    return estop_state;
+}
+
+void ThreadedEstopReader::readEstop()
+{
+    estop_msg = uart_reader->serialRead(ESTOP_MESSAGE_SIZE);
+
+    if (estop_msg.at(0) == 1)
+    {
+        estop_state = EstopState ::PLAY;
+    }
+    else if (estop_msg.at(0) == 0)
+    {
+        estop_state = EstopState::STOP;
+    }
+    else
+    {
+        estop_state = EstopState::STATUS_ERROR;
```

Comment:
you can probably just set this to `stop` or even `estop = true` w/ the log because there is nothing we can do here other than just log

specifically the wifi backend or whatever is owning this class can't do anything about a corrupt UART message detected here

Reply:
The idea here was that the owner of this class (estopManager #2027 ) would `log(fatal)` if we received a certain number of consecutive corrupt UART messages, and so it is useful to return the error state and not just play/stop

Reply:
another option would be to move the simple logic in estopManager to `threaded_estop_reader`, so that `threaded_estop_reader` provides the interface to the backend